frappe/public/js/frappe/form/grid_pagination.js

```diff
@@ -0,0 +1,132 @@
+
+		this.render_pagination();
+	}
+
+	render_pagination() {
+		if (this.grid.data.length <= this.page_length) {
+			this.wrapper.find('.grid-pagination').html('');
+		} else {
+			let $pagination_template = this.get_pagination_html();
+			this.wrapper.find('.grid-pagination').html($pagination_template);
+			this.prev_page_button = this.wrapper.find('.prev-page');
+			this.next_page_button = this.wrapper.find('.next-page');
+			this.$page_number = this.wrapper.find('.current-page-number');
+			this.$total_pages = this.wrapper.find('.total-page-number');
+			this.first_page_button = this.wrapper.find('.first-page');
+			this.last_page_button = this.wrapper.find('.last-page');
+
+			this.bind_pagination_events();
+		}
+	}
+
+	bind_pagination_events() {
+		this.prev_page_button.on('click', () => {
+			this.render_prev_page();
+		});
+
+		this.next_page_button.on('click', () => {
+			this.render_next_page();
+		});
+
+		this.first_page_button.on('click', () => {
+			this.go_to_page(1);
+		});
+
+		this.last_page_button.on('click', () => {
+			this.go_to_last_page();
+		});
+	}
+
+
+	update_page_numbers() {
+		let total_pages = Math.ceil(this.grid.data.length/this.page_length);
+		if (this.total_pages !== total_pages) {
+			this.total_pages = total_pages;
+			this.render_pagination();
+		}
+	}
+
+	check_page_number() {
+		if (this.page_index > this.total_pages && this.page_index > 1) {
+			this.go_to_page(this.page_index-1);
+		}
+	}
+
+	get_pagination_html() {
+		let page_text_html = `<div class="page-text">
+				<span class="current-page-number page-number">${__(this.page_index)}</span>
+				<span>${__('of')}</span>
+				<span class="total-page-number page-number"> ${__(this.total_pages)} </span> 
+			</div>`;
+
+		return $(`<button class="btn btn-default btn-xs first-page"">
+				<span class="first-page-icon">&laquo;</span>
+				<span>${__('First')}</span>
+			</button>
+			<a class="prev-page">&#8249;</a>
+			${page_text_html}
+			<a class="next-page">&#8250;</a>
+			<button class="btn btn-default btn-xs last-page">
+				<span>${__('Last')}</span>
+				<span class="first-page-icon">&raquo;</span>
+			</button>`);
+	}
+
+	render_next_page() {
+		if (this.page_index*this.page_length < this.grid.data.length) {
+			this.page_index++;
+			this.go_to_page();
+		}
+	}
+
+	render_prev_page() {
+		if (this.page_index > 1) {
+			this.page_index--;
+			this.go_to_page();
+		}
+	}
+
+	go_to_page(index) {
+		if (!index) {
+			index = this.page_index;
+		} else {
+			this.page_index = index;
+		}
+		let $rows = $(this.grid.parent).find(".rows").empty();
+		this.grid.render_result_rows($rows, true);
+		if (this.$page_number) {
+			this.$page_number.text(index);
+		}
+
+		this.update_page_numbers();
+	}
+
+	go_to_last_page() {
+		let total_pages = this.total_pages;
+		let page_length = this.page_length;
+		if (this.grid.data.length == page_length*total_pages) {
+			this.go_to_page(total_pages + 1);
```

Comment:
What's the logic of this line?

I guess it should be `this.grid.data.length > page_length * total_pages`

Reply:
This line was meant for adding rows to the last page. So that in case a row has needs to be added and `go_to_last_page` is called, if the total number of records is equal to the maximum number of records in `total_pages`, then it goes to the next page. So I've changed the function name to `go_to_last_page_to_add_row`